### ISSUE TITLE
exportSVG is way too slow

### ISSUE THREAD

**Issue by MuTsunTsai** [paperjs/paper.js#1895]
I noticed recently that for a project large enough, executing `exportSVG()` can be very slow. Sometimes it will even take more than 10 seconds just to complete. As I dig into the source code, I realized that the problem lies in line 294 of `src/svg/SvgExport.js`:

```
!parent || !Base.equals(parent[get](), value)
```

This part slows down the exporting significantly, mainly because the value of `parent[get]()` is not cached, so its value is re-computed for each child, and in each computation, it could iterate over all children again (as in the case of `getFillColor`); the accumulation of all these processes is simply wasting too much time.

I suppose that a complete fix would be to cache the value of `parent[get]()` throughout the course of `exportSVG()`, and the cache is cleared after exporting is finished.

Meanwhile, a quick fix would be simply replacing that entire expression with just `true` (i.e. don't care about parent value). The resulting SVG would be larger as a result, but the exporting speed is like a thousand times faster and it renders the same on screen.

I think a even better fix would be to add an option to `exportSVG()` that decides whether parent value should be compared; setting the option to `false` would make it behave the same way as replacing the expression by `true`.

**Comment by Spaxe** [comment]
Hi @MuTsunTsai could you make a PR to this issue since you forked it? It would be great as I'm running into this issue too.

It doesn't take much to hit the ceiling - something around ~300,000 elements will take forever to export.

**Comment by MuTsunTsai** [comment]
@Spaxe Well considering there's 49 pending PR already, I wouldn't count on that too much. Besides my fork is modified for my specific needs and is not for general purposes. I would suggest that you make similar modifications and make your own build as well.

**Comment by jokjus** [comment]
Thank you so, so much for this advice!! I've literally waited overnight for the svg export to finish… Now it's instantenous. Wow.

**Comment by MuTsunTsai** [comment]
@jokjus You're welcome; only that I've already abandoned paper.js partly because of the inactivity. If you're interested, I have an article talking about it: https://abstreamace.com/sglab/2023/03/12/bps-%E9%96%8B%E7%99%BC%E5%88%86%E4%BA%AB%E4%B9%8B-15%EF%BC%9A%E5%BE%9E-paper-%E5%88%B0-pixi/
It's in Chinese, but Google translate might work fine.

**Comment by lehni** [comment]
Thanks for the article @MuTsunTsai, and apologies for the silence here on this and many other issues. I understand your frustration too well. The sad reality is that maintaining open-source projects is an increasingly difficult burden that unfortunately hasn't worked out so well for Paper.js.

Ever since adding boolean operations, I've found myself spending all my available time to work on Paper.js on dealing with increasingly obscure edge cases, instead of more enjoyable and creative endeavours. Meanwhile, the JS ecosystem has changed so much that the code base is stuck in a state that would require a major rewrite before doing any bigger amounts of new work on it.

I still use Paper.js in my own work every now and then, and I still very much enjoy using it, hence the occasional releases, but I can't keep doing this free service to the community anymore, I simply don't find the time for it. I have so far not written an official announcement because I didn't know if this was a temporary or more permanent situation, and I am still not sure. But for the time being, I won't be able to put big amounts of work into solving bugs that I am not facing in my own work. I hope you understand.

**Comment by lehni** [comment]
Back to the reason of the slowdowns here, I think this is partly so slow because on groups, the style getters actually loop through all the children and unify the styles if all the children have the same style. This is very inefficient indeed, and should be cached / fixed.

**Comment by MuTsunTsai** [comment]
@lehni Yeah, I do understand that, and I appreciate your contribution to the community in any case. I have two suggestions though, if I may:
1. You really shouldn't attempt to develop your own vector boolean operation algorithm in the first place. This is an area that is well-developed in the literature and may fast and thoroughly tested algorithms exist, as mentioned in my article. You should just implement one of them instead of burn yourself out developing one.
2. If you find yourself unable to devote more into this project, talk to some of the major contributors and likely some of them will be willing to take over and take good care of the project. Don't just sit there and do nothing.

**Comment by lehni** [comment]
We didn't implement our own algorithm, we based our work on the work of the Skia team and their findings around the use of winding numbers, as outlined in this video: https://www.youtube.com/watch?v=OmfliNQsk88

The work was done in collaboration with @hkrish and @iconexperience and was actually a fun deep dive into geometric algorithms from which I learned a lot, and which is partly why I enjoy working on projects like Paper.js and Scriptographer.org before it. It's the maintenance aspect of it all that is less fun and motivating, as any long-term open-source maintainer will tell you. At one point your interests move on and you loose focus and interest. If by then the project has not become a source of income, it's very likely to fizzle out.

Our boolean operations might not be blazing fast, but they're actually pretty sturdy in lots of cases, and last time we checked, they compared quite well with Skia path ops, see here: https://github.com/paperjs/paper.js/issues/1619#issuecomment-444131740

As for the finding of contributors to take on the burden, of course I have been thinking about this, but the people who have been contenders in the past have also had similar situations as mine, and their focus has shifted, due to changes in circumstances, e.g. having children, work becoming more demanding, etc. Such is the story of life. If somebody wants to do take it on, they could also create an issue here and suggest doing so or start a discussion abot it. Would you be interested, for example?

Believe me, I have't been sitting there doing nothing. For the past year I was dealing with a demanding health situation, just one of multiple reasons why I haven't been able to find the time for Paper.js.

**Comment by MuTsunTsai** [comment]
@lehni In that case I'm afraid that you guys had been misled by Skia into thinking that their approach is THE one to follow. That approach is way slower than those algorithms based on sweep line (such as the MRF algorithm mentioned in my article). Indeed, one will have to study the paper, fill-in many missing implementation details, and figure out how to modify the algorithm to work with curves, but the result is like 100x faster. 

As for whether I would like to take over, should you ask me the same question a year ago, I will very likely say yes, but since I had moved on already, I'm afraid it's too late for that. 

And I apologize if my last comment didn't do you justice. I wish you all the best.

**Comment by MuTsunTsai** [comment]
And I wouldn't expect people to proactively request for project ownership transfer, by the way. I'm not sure about you guys, but at least in my culture it would be kinda rude to explicitly accuse the current owner for inactivity and demand handing over.

**Comment by lehni** [comment]
Haven't you done exactly that in your article though? Acuse me of inactivity?

**Comment by lehni** [comment]
We weren't misled by Skia, we were simply looking for a reasonable approach and thought we found one. We didn't find the one that you are using now, also polyBool.js didn't exist yet when we were looking around 2013.
You say it's 100x faster, but in your source-code, the stated performance gain over Paper.js is much lower, at 3x:

https://github.com/bp-studio/box-pleating-studio/blob/main/src/core/math/polyBool/README.md

<img width="1326" alt="image" src="https://github.com/paperjs/paper.js/assets/213275/592999f8-241c-40a8-b01a-76d94899de24">

Also, aren't you comparing apples with oranges here, because the algorithm is still only dealing with polygons in your use case, right? And it has been optimised for that use case?

There are other parts of your article that sound a bit bitter. For example you state that I rejected a WebGL backend right away, giving https://github.com/paperjs/paper.js/issues/917 as the reason. But this thread wasn't even about a WebGL drawing backend at all, it was about being able to use shaders inside the Paper.js canvas, and I didn't see an easy way to support that with the current 2D canvas based renderer at the time.  Now with transferable `ImageBitmap` objects in the API it's a different story, but the thread is from 2016, and back then the only way to do so would've been to rewrite the entire renderer then, and i didn't have the capacities to do so then. I never rejected the idea of an alternative WebGL based backend. I even suggested it to a friend who was working on 2D WebGL rendering at the time.

What I take away from this exchange with you, perhaps wrongly due to the translation to English of your article, is a tone of frustration and passive-aggressiveness towards Paper.js and me as one of its author and maintainer, which I think is rather sad. Paper.js helped you build something of your own, you learned how to structure a larger open-source project from it, and eventually decided it's time to move on, but without reaching out to me and suggesting possible ways of working together to make this actually work for you. At one point it worked so well for you that you "even wrote 'thanks to the author of paper.js' on the website", to quote your own words.

That's all good and totally your right to do of course, but why the hostility now? We've given you something for free, and you were able to build something else on top of it and let it grow, eventually outgrowing the capabilities of the foundation, as it so often happens. Do you really need to dump on it as you decide to move on?

**Comment by MuTsunTsai** [comment]
@lehni I was merely stating fact, without blaming you at all and without any demand.

**Comment by MuTsunTsai** [comment]
The statement about the 3x performance gain was before I re-implement MRF algorithm myself (as mentioned in the article, polybooljs actually did a poor job on that). And after that, it's pretty much 100x faster indeed. 

And nope, like I said, it's been modified to handle any curved shapes as well. So I'm not comparing apples to oranges.

You said I moved on without reaching out to you. I wonder why this issue itself doesn't count? You made absolutely no reply to the issue for two years, and suddenly decided to make it sound like my fault soon after saying “you understand my frustration”? Seriously?

As a matter of fact I think you're the hostile one at this moment, not me. I was merely giving feedbacks. Maybe I should just remain silent like you did.

**Comment by lehni** [comment]
I won't engage in this exchange further and wish you the best of luck with your endeavours. Maybe one day after maintaining an open-source project with a whole lot of users for a couple of years you will understand why not answering to every single one of over 300 open issues isn't automatically hostile or personal.

**Comment by MuTsunTsai** [comment]
Of course, but don't say I didn't reach out to you because I totally did.

**Comment by lehni** [comment]
<img width="534" alt="image" src="https://github.com/paperjs/paper.js/assets/213275/d9ee49e1-feea-41d8-b42f-61df8112cb83">

**Comment by MuTsunTsai** [comment]
Should I be impressed by tags?

**Comment by lehni** [comment]
Back to the reason of the slowdowns here, I think this is partly so slow because on groups, the style getters actually loop through all the children and unify the styles if all the children have the same style. This is very inefficient indeed, and should be cached / fixed.

**Comment by MuTsunTsai** [comment]
And thanks for the heads-up, only that I, unlike you, will be sure to leave a note on the repo about how busy I am.

**Comment by lehni** [comment]
I've added `options.reduceAttributes` to control this behavior now. You can explicitly set it to false for faster output, but it will contain a lot of redundant attributes
